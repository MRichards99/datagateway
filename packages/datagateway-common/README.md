# datagateway-common

A library containint common code and components for the DataGateway project.
It is a dependency of other packages in the repo and other apps can import code from it
like normal.

When adding new components/functions/definitions/constants etc., you must export the item
in `index.tsx` so that it will be importable from other applications.

e.g.

In `some_file.tsx`

```
export const my_const = 1;
```

In `index.tsx`

```
export { my_const } from 'some_file.tsx'
export { my_const };
```

If you export a lot of items in `some_file.tsx` and want to export all of them to be exposed to other packages
then the following syntax can be used instead

`index.tsx`

```
export * from 'some_file.tsx'
```

## Available Scripts

In the project directory, you can run:

### `npm install`

Installs dependencies for `datagateway-common` only. Should normally use `npx lerna add` at the
top level, but can use regular `npm install` if it's only used by `datagateway-common` and not other
packages.

### `npm start`

Runs the app in development mode
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

In order to begin local development, you should uncomment the lines in `index.tsx`.
These allow the app to load like a regular React app rather than behave as a library.

### `npm run tsc`

Runs the Typescript compiler that converts our Typescript files into a Javascript library
that can be imported by our other packages. When making changes to this app, you must run
this to see the changes reflected in other apps. Think of it as a build step like `npm run build`.

### `npm test`

Runs unit tests

### `npm run test:watch`

Runs unit tests in watch mode, allowing you to filter which tests to run and automatically restart tests on code changes.

### `npm run lint:js`

Lints the `src` folder
