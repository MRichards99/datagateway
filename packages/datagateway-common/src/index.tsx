// Uncomment to use App to test components
// import React from 'react';
// import ReactDOM from 'react-dom';
// import App from './App';

export * from './app.types';
export * from './state/actions/actions.types';

export { default as Table } from './table/table.component';
export * from './table/table.component';
export { default as DetailsPanelRow } from './table/rowRenderers/detailsPanelRow.component';
export { default as DataHeader } from './table/headerRenderers/dataHeader.component';
export { default as TextColumnFilter } from './table/columnFilters/textColumnFilter.component';
export { default as DateColumnFilter } from './table/columnFilters/dateColumnFilter.component';
export { default as ActionCellComponent } from './table/cellRenderers/actionCell.component';
export { default as DataCellComponent } from './table/cellRenderers/dataCell.component';
export { default as ExpandCellComponent } from './table/cellRenderers/expandCell.component';
export * from './table/cellRenderers/cellContentRenderers';
export * from './state/actions/datafiles';
export * from './state/actions/datasets';
export * from './state/actions/investigations';
export * from './state/actions/cart';
export * from './state/actions/instruments';
export * from './state/actions/facilityCycles';
export { default as Preloader } from './preloader/preloader.component';
export { default as dGCommonReducer } from './state/reducers/dgcommon.reducer';

import { StateType } from './state/app.types';
export type DGCommonState = StateType;

// ReactDOM.render(<App />, document.getElementById('root'));
