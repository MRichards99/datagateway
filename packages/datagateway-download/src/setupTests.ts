import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

// below required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
function noOp(): void { }

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp });
}

export const flushPromises = (): Promise<void> => new Promise(setImmediate);

// Mock lodash.debounce to return the function we want to call.
jest.mock('lodash.debounce', () => (fn: Function) => fn);
