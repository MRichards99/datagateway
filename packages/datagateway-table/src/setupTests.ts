import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Action } from 'redux';
import { StateType } from './state/app.types';
import { initialState } from './state/reducers/dgtable.reducer';
// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

// below required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
function noOp(): void {}

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp });
}

// these are used for testing async actions
export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getState = (): Partial<StateType> => ({ dgtable: initialState });
export const dispatch = (action: Action): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getState);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

// Mock lodash.debounce to return the function we want to call.
jest.mock('lodash.debounce', () => (fn: Function) => fn);
