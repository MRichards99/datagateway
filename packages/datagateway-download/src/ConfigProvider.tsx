import React from 'react';
import axios from 'axios';

import * as log from 'loglevel';
import { Preloader } from 'datagateway-common';

export interface DownloadSettings {
  facilityName: string;
  apiUrl: string;
  downloadApiUrl: string;
  idsUrl: string;
  accessMethods: {
    type: string;
    idsUrl: string;
    displayName?: string;
    description?: string;
  }[];
}

const initialConfiguration = {
  facilityName: '',
  apiUrl: '',
  downloadApiUrl: '',
  idsUrl: '',
  accessMethods: [],
};

export const DownloadSettingsContext = React.createContext<DownloadSettings>(
  initialConfiguration
);

class ConfigProvider extends React.Component<
  { children: React.ReactNode },
  { loading: boolean; settings: DownloadSettings }
> {
  public constructor(props: { children: React.ReactNode }) {
    super(props);

    this.state = {
      loading: true,
      settings: initialConfiguration,
    };
  }

  public componentDidMount(): void {
    this.updateConfigurationState();
  }

  private updateConfigurationState = async () => {
    // let updatedState = this.state;
    const settings = await axios
      .get<DownloadSettings>('/datagateway-download-settings.json')
      .then(res => {
        const settings = res.data;
        console.log('Download settings: ', settings);

        if (typeof settings !== 'object') {
          throw Error('Invalid format');
        }

        // Ensure the facility name exists.
        if (!('facilityName' in settings)) {
          throw new Error('facilityName is undefined in settings');
        }

        // Ensure all API related URLs are present.
        if (
          !(
            'idsUrl' in settings &&
            'apiUrl' in settings &&
            'downloadApiUrl' in settings
          )
        ) {
          throw new Error(
            'One of the URL options (idsUrl, apiUrl, downloadApiUrl) is undefined in settings'
          );
        }

        // Ensure access methods are present in the configuration.
        if (!('accessMethods' in settings)) {
          throw new Error('accessMethods is undefined in settings');
        } else {
          // Check to ensure at least one access method has been defined.
          if (settings['accessMethods'].length < 1) {
            throw new Error(
              'At least one access method should be defined under accessMethods in settings'
            );
          } else {
            // Check all defined access methods to ensure type and idsUrl have been stated.
            for (let i = 0; i < settings['accessMethods'].length; i++) {
              const method = settings['accessMethods'][i];
              if (!(method['type'] && method['idsUrl']))
                throw new Error(
                  `An access method (${i}) defined in settings does not contain type or idsUrl`
                );
            }
          }
        }

        return settings;
      })
      .catch(error => {
        log.error(
          `Error loading datagateway-download-settings.json: ${error.message}`
        );
        return null;
      });

    if (settings !== null) {
      // updatedState = {
      //   loading: false,
      //   settings: settings,
      // };
      this.setState({
        loading: false,
        settings: settings,
      });
    }
  };

  public render(): React.ReactElement {
    return (
      // We pass the download settings that has been loaded
      // for all child components to consume, if required.
      <Preloader loading={this.state.loading}>
        <DownloadSettingsContext.Provider value={this.state.settings}>
          {this.props.children}
        </DownloadSettingsContext.Provider>
      </Preloader>
    );
  }
}

export const ConfigConsumer = DownloadSettingsContext.Consumer;
export default ConfigProvider;
