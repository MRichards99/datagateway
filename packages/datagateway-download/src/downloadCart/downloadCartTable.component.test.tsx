import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import DownloadCartTable from './downloadCartTable.component';
import { DownloadCartItem } from 'datagateway-common';
import { flushPromises } from '../setupTests';
import {
  fetchDownloadCartItems,
  removeAllDownloadCartItems,
  removeDownloadCartItem,
  getSize,
  getCartDatafileCount,
} from '../downloadApi';
import { act } from 'react-dom/test-utils';
import { DownloadSettingsContext } from '../ConfigProvider';

jest.mock('../downloadApi');
jest.useFakeTimers();

describe('Download cart table component', () => {
  let shallow;
  let mount;

  const cartItems: DownloadCartItem[] = [
    {
      entityId: 1,
      entityType: 'investigation',
      id: 1,
      name: 'INVESTIGATION 1',
      parentEntities: [],
    },
    {
      entityId: 2,
      entityType: 'investigation',
      id: 2,
      name: 'INVESTIGATION 2',
      parentEntities: [],
    },
    {
      entityId: 3,
      entityType: 'dataset',
      id: 3,
      name: 'DATASET 1',
      parentEntities: [],
    },
    {
      entityId: 4,
      entityType: 'datafile',
      id: 4,
      name: 'DATAFILE 1',
      parentEntities: [],
    },
  ];

  (fetchDownloadCartItems as jest.Mock).mockImplementation(() =>
    Promise.resolve(cartItems)
  );

  (removeAllDownloadCartItems as jest.Mock).mockImplementation(() =>
    Promise.resolve()
  );

  (removeDownloadCartItem as jest.Mock).mockImplementation(() =>
    Promise.resolve()
  );

  (getSize as jest.Mock).mockImplementation(() => Promise.resolve(1));

  (getCartDatafileCount as jest.Mock).mockImplementation(() =>
    Promise.resolve(7)
  );

  // Create our mocked datagateway-download settings file.
  const mockedSettings = {
    facilityName: 'LILS',
    apiUrl: 'http://scigateway-preprod.esc.rl.ac.uk:5000',
    downloadApiUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/topcat',
    idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
    accessMethods: {
      https: {
        idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
        displayName: 'HTTPS',
        description: 'Example description for HTTPS access method.',
      },
      globus: {
        idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
        displayName: 'Globus',
        description: 'Example description for Globus access method.',
      },
    },
  };

  beforeEach(() => {
    shallow = createShallow({ untilSelector: 'div' });
    mount = createMount();
  });

  afterEach(() => {
    mount.cleanUp();
    (fetchDownloadCartItems as jest.Mock).mockClear();
    (getSize as jest.Mock).mockClear();
    (getCartDatafileCount as jest.Mock).mockClear();
    (removeAllDownloadCartItems as jest.Mock).mockClear();
    (removeDownloadCartItem as jest.Mock).mockClear();
    jest.clearAllTimers();
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <DownloadCartTable statusTabRedirect={jest.fn()} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('fetches the download cart on load', async () => {
    const wrapper = mount(
      <DownloadSettingsContext.Provider value={mockedSettings}>
        <DownloadCartTable statusTabRedirect={jest.fn()} />
      </DownloadSettingsContext.Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(fetchDownloadCartItems).toHaveBeenCalled();
  });

  it('calculates sizes once cart items have been fetched', async () => {
    const wrapper = mount(
      <DownloadSettingsContext.Provider value={mockedSettings}>
        <DownloadCartTable statusTabRedirect={jest.fn()} />
      </DownloadSettingsContext.Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(getSize).toHaveBeenCalledTimes(4);
    expect(
      wrapper
        .find('[aria-colindex=3]')
        .find('p')
        .first()
        .text()
    ).toEqual('1 B');
    expect(wrapper.find('p#totalSizeDisplay').text()).toEqual(
      expect.stringContaining('Total size: 4 B')
    );
  });

  it('calculates total file count once cart items have been fetched', async () => {
    const wrapper = mount(
      <DownloadSettingsContext.Provider value={mockedSettings}>
        <DownloadCartTable statusTabRedirect={jest.fn()} />
      </DownloadSettingsContext.Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(getCartDatafileCount).toHaveBeenCalled();
    expect(wrapper.find('p#fileCountDisplay').text()).toEqual(
      expect.stringContaining('Number of files: 7')
    );
  });

  // TODO: If I place a axios response here for status call, why is it undefined?
  it('loads cart confirmation dialog when Download Cart button is clicked', async () => {
    const wrapper = mount(
      <DownloadSettingsContext.Provider value={mockedSettings}>
        <DownloadCartTable statusTabRedirect={jest.fn()} />
      </DownloadSettingsContext.Provider>
    );

    expect(wrapper.find('button#downloadCartButton').prop('disabled')).toBe(
      true
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(wrapper.find('button#downloadCartButton').prop('disabled')).toBe(
      false
    );

    wrapper.find('button#downloadCartButton').simulate('click');

    // Update the wrapper with the loading dialog.
    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    // Update the wrapper with the download confirmation dialog.
    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(wrapper.exists('[aria-labelledby="downloadCartConfirmation"]')).toBe(
      true
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    // Close the confirmation dialog.
    wrapper
      .find('button[aria-label="download-confirmation-close"]')
      .simulate('click');
  });

  it('calls clearCart function once the download cart is closed', async () => {});

  it('removes all items from cart when Remove All button is clicked', async () => {
    const wrapper = mount(
      <DownloadSettingsContext.Provider value={mockedSettings}>
        <DownloadCartTable statusTabRedirect={jest.fn()} />
      </DownloadSettingsContext.Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    await act(async () => {
      wrapper.find('button#removeAllButton').simulate('click');
      await flushPromises();
      wrapper.update();
    });

    expect(removeAllDownloadCartItems).toHaveBeenCalled();
    expect(wrapper.exists('[role="gridcell"]')).toBe(false);
    expect(wrapper.exists('[aria-rowcount=0]')).toBe(true);
  });

  it("removes an item when said item's remove button is clicked", async () => {
    const wrapper = mount(
      <DownloadSettingsContext.Provider value={mockedSettings}>
        <DownloadCartTable statusTabRedirect={jest.fn()} />
      </DownloadSettingsContext.Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    wrapper
      .find('button[aria-label="Remove INVESTIGATION 2 from cart"]')
      .simulate('click');

    expect(
      wrapper
        .find('button[aria-label="Remove INVESTIGATION 2 from cart"] svg')
        .parent()
        .prop('color')
    ).toEqual('error');

    await act(async () => {
      jest.runAllTimers();
      await flushPromises();
      wrapper.update();
    });

    expect(removeDownloadCartItem).toHaveBeenCalled();
    expect(removeDownloadCartItem).toHaveBeenCalledWith(2, 'investigation', {
      facilityName: mockedSettings.facilityName,
      downloadApiUrl: mockedSettings.downloadApiUrl,
    });
    expect(
      wrapper.exists('[aria-label="Remove INVESTIGATION 2 from cart"]')
    ).toBe(false);
    expect(wrapper.exists('[aria-rowcount=3]')).toBe(true);
  });

  it('sorts data when headers are clicked', async () => {
    const wrapper = mount(
      <DownloadSettingsContext.Provider value={mockedSettings}>
        <DownloadCartTable statusTabRedirect={jest.fn()} />
      </DownloadSettingsContext.Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    const firstNameCell = wrapper
      .find('[aria-colindex=1]')
      .find('p')
      .first();

    const typeSortLabel = wrapper
      .find('[role="columnheader"] span[role="button"]')
      .at(1);

    typeSortLabel.simulate('click');

    expect(firstNameCell.text()).toEqual('DATAFILE 1');

    typeSortLabel.simulate('click');

    expect(firstNameCell.text()).toEqual('INVESTIGATION 1');

    const nameSortLabel = wrapper
      .find('[role="columnheader"] span[role="button"]')
      .at(0);

    nameSortLabel.simulate('click');

    expect(firstNameCell.text()).toEqual('INVESTIGATION 1');

    nameSortLabel.simulate('click');

    expect(firstNameCell.text()).toEqual('INVESTIGATION 2');

    nameSortLabel.simulate('click');

    expect(firstNameCell.text()).toEqual('INVESTIGATION 1');
  });

  it('filters data when text fields are typed into', async () => {
    const wrapper = mount(
      <DownloadSettingsContext.Provider value={mockedSettings}>
        <DownloadCartTable statusTabRedirect={jest.fn()} />
      </DownloadSettingsContext.Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    const nameFilterInput = wrapper.find('[aria-label="Filter by Name"] input');
    nameFilterInput.instance().value = '1';
    nameFilterInput.simulate('change');

    expect(wrapper.exists('[aria-rowcount=3]')).toBe(true);
    expect(
      wrapper.exists('[aria-label="Remove INVESTIGATION 2 from cart"]')
    ).toBe(false);

    const typeFilterInput = wrapper.find('[aria-label="Filter by Type"] input');
    typeFilterInput.instance().value = 'data';
    typeFilterInput.simulate('change');

    expect(wrapper.exists('[aria-rowcount=2]')).toBe(true);
    expect(
      wrapper.exists('[aria-label="Remove INVESTIGATION 1 from cart"]')
    ).toBe(false);

    typeFilterInput.instance().value = '';
    typeFilterInput.simulate('change');

    expect(wrapper.exists('[aria-rowcount=3]')).toBe(true);
    expect(
      wrapper.exists('[aria-label="Remove INVESTIGATION 1 from cart"]')
    ).toBe(true);

    nameFilterInput.instance().value = '';
    nameFilterInput.simulate('change');

    expect(wrapper.exists('[aria-rowcount=4]')).toBe(true);
    expect(
      wrapper.exists('[aria-label="Remove INVESTIGATION 2 from cart"]')
    ).toBe(true);
  });
});
