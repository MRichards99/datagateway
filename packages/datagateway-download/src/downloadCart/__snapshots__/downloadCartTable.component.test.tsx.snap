// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Download cart table component renders correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Grid))
    container={true}
    direction="column"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <WithStyles(ForwardRef(Paper))
        style={
          Object {
            "height": "calc(100vh - 150px)",
          }
        }
      >
        <WithStyles(VirtualizedTable)
          actions={
            Array [
              [Function],
            ]
          }
          columns={
            Array [
              Object {
                "dataKey": "name",
                "filterComponent": [Function],
                "label": "Name",
              },
              Object {
                "dataKey": "entityType",
                "filterComponent": [Function],
                "label": "Type",
              },
              Object {
                "cellContentRenderer": [Function],
                "dataKey": "size",
                "label": "Size",
              },
            ]
          }
          data={Array []}
          loading={true}
          onSort={[Function]}
          sort={Object {}}
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      alignItems="flex-end"
      container={true}
      direction="column"
      item={true}
      justify="space-between"
    >
      <WithStyles(ForwardRef(Grid))
        alignContent="flex-end"
        container={true}
        direction="column"
        item={true}
        style={
          Object {
            "marginRight": "1.2em",
          }
        }
        xs={3}
      >
        <WithStyles(ForwardRef(Typography))
          id="fileCountDisplay"
        >
          Number of files: 
          Calculating...
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          id="totalSizeDisplay"
        >
          Total size:
           
          0 B
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        container={true}
        item={true}
        justify="flex-end"
        spacing={1}
        style={
          Object {
            "marginRight": "1em",
          }
        }
        xs={3}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            id="removeAllButton"
            onClick={[Function]}
            variant="contained"
          >
            Remove All
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            disabled={true}
            id="downloadCartButton"
            onClick={[Function]}
            variant="contained"
          >
            Download Cart
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(DownloadConfirmDialog)
    aria-labelledby="downloadCartConfirmation"
    isTwoLevel={false}
    setClose={[Function]}
    setOpen={false}
    totalSize={0}
  />
</div>
`;
